### PR TITLE
perf(consumer): Only stage offsets when ready to commit

### PR DESCRIPTION
**Pull request by tkaemming**
Just a minor (potential) performance improvement that may be worth doing that I noticed while I was moving other things around. I vaguely recall @fpacifici mentioning that `stage_offsets` was a surprisingly substantial part of the work done by the `_handle_message` method in some earlier performance profiling. I did not validate that (so I can't say how big the actual improvement is, if any) but it seemed like an easy enough improvement to amortize the potential cost of this staging offsets over the entire batch.

### CODE REVIEW

File: snuba/utils/streams/batching.py [hunk 1]
```diff
@@ -247,14 +245,17 @@ def _flush(self, force: bool = False) -> None:
                 "batch.flush.normalized", flush_duration / batch_results_length
             )
 
-        logger.debug("Committing offsets")
+        logger.debug("Committing offsets for batch")
         commit_start = time.time()
-        self._commit()
+        self.consumer.stage_offsets(
+            {
+                partition: offsets.hi + 1
```
Comment (tkaemming) [inline snuba/utils/streams/batching.py:252]: This is kind of kludgy, alternatively the `Offsets` structure could represent the lower bound inclusive and upper bound exclusive (like slice syntax) which might make more sense anyway.

Comment (fpacifici) [reply snuba/utils/streams/batching.py:252]: What's the impact of changing the Offset structure ? (read: how many things do we expect to silently break :D ? )

**Review by fpacifici — APPROVED** [review]
Up to you if you want to make the change to the Offset structure. Not sure how confident we are on our test coverage.

One remark about the expected performance gain. I measured that on a cdc consumer that was doing almost nothing other than writing stats and checking message keys. Not sure how this will be impactful on events/errors/transactions consumers. Still certainly worth.

File: snuba/utils/streams/batching.py (hunk 1 above)
Comment (tkaemming) [reply snuba/utils/streams/batching.py:252]: It is only used in `BatchingConsumer` to my knowledge (and its usage is "private" within that) so hopefully nothing.

**Review by tkaemming — COMMENTED** [review]
> Up to you if you want to make the change to the Offset structure. Not sure how confident we are on our test coverage.

There is test coverage here, though it's not very thorough (it ensures that a commit occurs, but doesn't validate the commit state itself.) I'll go ahead and make that change, mostly since the arithmetic seems kind of magic number-y when we have a perfectly good `get_next_offset` method on `Message`.

> One remark about the expected performance gain. I measured that on a cdc consumer that was doing almost nothing other than writing stats and checking message keys. Not sure how this will be impactful on events/errors/transactions consumers. Still certainly worth.

Ahh, right — good point, thanks for the context.

**Pull request merged**